# Preprocessing the Titanic Dataset

## import the relevant libraries

In [1]:
import pandas as pd

## load the Titanic dataset

In [2]:
raw_csv_data = pd.read_csv('Titanic_raw_data.csv')

In [3]:
df = raw_csv_data.copy()
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## dropping non critical columns that can be detected logically

### we will drop the 'name' column because it does not has any predictive power

In [4]:
df = df.drop(['Name'], axis = 1)
df.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0            1         0       3    male  22.0      1      0   
1            2         1       1  female  38.0      1      0   
2            3         1       3  female  26.0      0      0   
3            4         1       1  female  35.0      1      0   
4            5         0       3    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

### we will drop the 'id' column because it doesn't have any predictive power and already the rows are ordered from 0 to 890

In [5]:
df = df.drop(['PassengerId'], axis = 1)
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch            Ticket     Fare  \
0         0       3    male  22.0      1      0         A/5 21171   7.2500   
1         1       1  female  38.0      1      0          PC 17599  71.2833   
2         1       3  female  26.0      0      0  STON/O2. 3101282   7.9250   
3         1       1  female  35.0      1      0            113803  53.1000   
4         0       3    male  35.0      0      0            373450   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

### we will drop the Ticket columns because the fare column will show us the economic level of the passenger and it's enough

In [6]:
df = df.drop(['Ticket'], axis = 1)
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22.0      1      0   7.2500   NaN        S
1         1       1  female  38.0      1      0  71.2833   C85        C
2         1       3  female  26.0      0      0   7.9250   NaN        S
3         1       1  female  35.0      1      0  53.1000  C123        S
4         0       3    male  35.0      0      0   8.0500   NaN        S

## checking for missing values

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


### Cabin, Age and Embarked has missing values, we will fill the NaN in Cabin with unknown instead and then create a Deck column to display just the first letter because we don't need to know the number it's enough to know the cabin|

In [8]:
df['Cabin'].fillna('Unknown', inplace=True)
df['Deck'] = df['Cabin'].str[0]
df = df.drop(['Cabin'], axis = 1)
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked Deck
0           0       3    male  22.0      1      0   7.2500        S    U
1           1       1  female  38.0      1      0  71.2833        C    C
2           1       3  female  26.0      0      0   7.9250        S    U
3           1       1  female  35.0      1      0  53.1000        S    C
4           0       3    male  35.0      0      0   8.0500        S    U
..        ...     ...     ...   ...    ...    ...      ...      ...  ...
886         0       2    male  27.0      0      0  13.0000        S    U
887         1       1  female  19.0      0      0  30.0000        S    B
888         0       3  female   NaN      1      2  23.4500        S    U
889         1       1    male  26.0      0      0  30.0000        C    C
890         0       3    male  32.0      0      0   7.7500        Q    U

[891 rows x 9 columns]

### since embarked has just 2 missing values we can fill it with any of it's values this will not affect the analysis

In [9]:
df['Embarked'].fillna('Q', inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  891 non-null    object 
 8   Deck      891 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


### we will replace the NaN values in Age with the mean value

In [10]:
mean_age = df['Age'].mean()
df['Age'].fillna(mean_age, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       891 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Embarked  891 non-null    object 
 8   Deck      891 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


### now no missing values are shown, we will just put the 'Survived' column at the end because it will be our target

In [11]:
columns_ordered = ['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked', 'Deck', 'Survived']
df_ordered = df[columns_ordered]
df_ordered.head()

Pclass     Sex   Age  SibSp  Parch     Fare Embarked Deck  Survived
0       3    male  22.0      1      0   7.2500        S    U         0
1       1  female  38.0      1      0  71.2833        C    C         1
2       3  female  26.0      0      0   7.9250        S    U         1
3       1  female  35.0      1      0  53.1000        S    C         1
4       3    male  35.0      0      0   8.0500        S    U         0

## Sex

### we will replace male with 1's and female with 0's is the Sex column

In [12]:
df_ordered['Sex'] = df_ordered['Sex'].replace({'male': 1, 'female': 0})
df_ordered.head()

Pclass  Sex   Age  SibSp  Parch     Fare Embarked Deck  Survived
0       3    1  22.0      1      0   7.2500        S    U         0
1       1    0  38.0      1      0  71.2833        C    C         1
2       3    0  26.0      0      0   7.9250        S    U         1
3       1    0  35.0      1      0  53.1000        S    C         1
4       3    1  35.0      0      0   8.0500        S    U         0

## Age

In [13]:
distinct_ages = df_ordered['Age'].unique()
distinct_ages

array([22.        , 38.        , 26.        , 35.        , 29.69911765,
       54.        ,  2.        , 27.        , 14.        ,  4.        ,
       58.        , 20.        , 39.        , 55.        , 31.        ,
       34.        , 15.        , 28.        ,  8.        , 19.        ,
       40.        , 66.        , 42.        , 21.        , 18.        ,
        3.        ,  7.        , 49.        , 29.        , 65.        ,
       28.5       ,  5.        , 11.        , 45.        , 17.        ,
       32.        , 16.        , 25.        ,  0.83      , 30.        ,
       33.        , 23.        , 24.        , 46.        , 59.        ,
       71.        , 37.        , 47.        , 14.5       , 70.5       ,
       32.5       , 12.        ,  9.        , 36.5       , 51.        ,
       55.5       , 40.5       , 44.        ,  1.        , 61.        ,
       56.        , 50.        , 36.        , 45.5       , 20.5       ,
       62.        , 41.        , 52.        , 63.        , 23.5 

### we can see that there is fractional values in the age, but it's better to have them as discrete values in term of memory usage and performance

In [14]:
df_ordered['Age'] = df_ordered['Age'].astype(int)
distinct_ages = df_ordered['Age'].unique()
distinct_ages

array([22, 38, 26, 35, 29, 54,  2, 27, 14,  4, 58, 20, 39, 55, 31, 34, 15,
       28,  8, 19, 40, 66, 42, 21, 18,  3,  7, 49, 65,  5, 11, 45, 17, 32,
       16, 25,  0, 30, 33, 23, 24, 46, 59, 71, 37, 47, 70, 12,  9, 36, 51,
       44,  1, 61, 56, 50, 62, 41, 52, 63, 43, 60, 10, 64, 13, 48, 53, 57,
       80,  6, 74])

### all ages converted sucessfully

In [15]:
df_ordered.head()

Pclass  Sex  Age  SibSp  Parch     Fare Embarked Deck  Survived
0       3    1   22      1      0   7.2500        S    U         0
1       1    0   38      1      0  71.2833        C    C         1
2       3    0   26      0      0   7.9250        S    U         1
3       1    0   35      1      0  53.1000        S    C         1
4       3    1   35      0      0   8.0500        S    U         0

## Embarked

In [16]:
Embarked_disticnt_values = df_ordered['Embarked'].unique()
Embarked_disticnt_values

array(['S', 'C', 'Q'], dtype=object)

In [17]:
Embarked_columns = pd.get_dummies(df_ordered['Embarked'])
Embarked_columns = Embarked_columns.astype(int)
Embarked_columns

C  Q  S
0    0  0  1
1    1  0  0
2    0  0  1
3    0  0  1
4    0  0  1
..  .. .. ..
886  0  0  1
887  0  0  1
888  0  0  1
889  1  0  0
890  0  1  0

[891 rows x 3 columns]

In [18]:
df_ordered = pd.concat([df_ordered, Embarked_columns], axis = 1)
df_ordered = df_ordered.drop(['Embarked'], axis = 1)

df_ordered.head()

Pclass  Sex  Age  SibSp  Parch     Fare Deck  Survived  C  Q  S
0       3    1   22      1      0   7.2500    U         0  0  0  1
1       1    0   38      1      0  71.2833    C         1  1  0  0
2       3    0   26      0      0   7.9250    U         1  0  0  1
3       1    0   35      1      0  53.1000    C         1  0  0  1
4       3    1   35      0      0   8.0500    U         0  0  0  1

### we will replace Embarked letters with teh full port name to avoid repetition when creating dummies for Deck and to be more clear

In [20]:
columns_ordered = ['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
        'Deck', 'Survived', 'Cherbourg', 'Queenstown', 'Southampton']
df_ordered.columns = columns_ordered
df_ordered.head()

Pclass  Sex  Age  SibSp  Parch     Fare Deck  Survived  Cherbourg  \
0       3    1   22      1      0   7.2500    U         0          0   
1       1    0   38      1      0  71.2833    C         1          1   
2       3    0   26      0      0   7.9250    U         1          0   
3       1    0   35      1      0  53.1000    C         1          0   
4       3    1   35      0      0   8.0500    U         0          0   

   Queenstown  Southampton  
0           0            1  
1           0            0  
2           0            1  
3           0            1  
4           0            1

## Deck

In [21]:
Deck_distinct_values = df_ordered['Deck'].unique()
Deck_distinct_values

array(['U', 'C', 'E', 'G', 'D', 'A', 'B', 'F', 'T'], dtype=object)

In [22]:
Deck_columns = pd.get_dummies(df_ordered['Deck'])
Deck_columns = Deck_columns.astype(int)
Deck_columns

A  B  C  D  E  F  G  T  U
0    0  0  0  0  0  0  0  0  1
1    0  0  1  0  0  0  0  0  0
2    0  0  0  0  0  0  0  0  1
3    0  0  1  0  0  0  0  0  0
4    0  0  0  0  0  0  0  0  1
..  .. .. .. .. .. .. .. .. ..
886  0  0  0  0  0  0  0  0  1
887  0  1  0  0  0  0  0  0  0
888  0  0  0  0  0  0  0  0  1
889  0  0  1  0  0  0  0  0  0
890  0  0  0  0  0  0  0  0  1

[891 rows x 9 columns]

In [23]:
df_ordered = pd.concat([df_ordered, Deck_columns], axis = 1)
df_ordered = df_ordered.drop(['Deck'], axis = 1)

df_ordered.head()

Pclass  Sex  Age  SibSp  Parch     Fare  Survived  Cherbourg  Queenstown  \
0       3    1   22      1      0   7.2500         0          0           0   
1       1    0   38      1      0  71.2833         1          1           0   
2       3    0   26      0      0   7.9250         1          0           0   
3       1    0   35      1      0  53.1000         1          0           0   
4       3    1   35      0      0   8.0500         0          0           0   

   Southampton  A  B  C  D  E  F  G  T  U  
0            1  0  0  0  0  0  0  0  0  1  
1            0  0  0  1  0  0  0  0  0  0  
2            1  0  0  0  0  0  0  0  0  1  
3            1  0  0  1  0  0  0  0  0  0  
4            1  0  0  0  0  0  0  0  0  1

In [24]:
columns_ordered = ['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
         'Cherbourg', 'Queenstown', 'Southampton', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T', 'U', 'Survived']
df_ordered = df_ordered[columns_ordered]
df_ordered.head()

Pclass  Sex  Age  SibSp  Parch     Fare  Cherbourg  Queenstown  \
0       3    1   22      1      0   7.2500          0           0   
1       1    0   38      1      0  71.2833          1           0   
2       3    0   26      0      0   7.9250          0           0   
3       1    0   35      1      0  53.1000          0           0   
4       3    1   35      0      0   8.0500          0           0   

   Southampton  A  B  C  D  E  F  G  T  U  Survived  
0            1  0  0  0  0  0  0  0  0  1         0  
1            0  0  0  1  0  0  0  0  0  0         1  
2            1  0  0  0  0  0  0  0  0  1         1  
3            1  0  0  1  0  0  0  0  0  0         1  
4            1  0  0  0  0  0  0  0  0  1         0

### making sure of the types of the features

In [25]:
df_ordered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Pclass       891 non-null    int64  
 1   Sex          891 non-null    int64  
 2   Age          891 non-null    int32  
 3   SibSp        891 non-null    int64  
 4   Parch        891 non-null    int64  
 5   Fare         891 non-null    float64
 6   Cherbourg    891 non-null    int32  
 7   Queenstown   891 non-null    int32  
 8   Southampton  891 non-null    int32  
 9   A            891 non-null    int32  
 10  B            891 non-null    int32  
 11  C            891 non-null    int32  
 12  D            891 non-null    int32  
 13  E            891 non-null    int32  
 14  F            891 non-null    int32  
 15  G            891 non-null    int32  
 16  T            891 non-null    int32  
 17  U            891 non-null    int32  
 18  Survived     891 non-null    int64  
dtypes: float

### Saving the PreProcessed dataset into a *.csv file

In [26]:
df_preprocessed = df_ordered.copy()
df_preprocessed.head(20)

Pclass  Sex  Age  SibSp  Parch     Fare  Cherbourg  Queenstown  \
0        3    1   22      1      0   7.2500          0           0   
1        1    0   38      1      0  71.2833          1           0   
2        3    0   26      0      0   7.9250          0           0   
3        1    0   35      1      0  53.1000          0           0   
4        3    1   35      0      0   8.0500          0           0   
5        3    1   29      0      0   8.4583          0           1   
6        1    1   54      0      0  51.8625          0           0   
7        3    1    2      3      1  21.0750          0           0   
8        3    0   27      0      2  11.1333          0           0   
9        2    0   14      1      0  30.0708          1           0   
10       3    0    4      1      1  16.7000          0           0   
11       1    0   58      0      0  26.5500          0           0   
12       3    1   20      0      0   8.0500          0           0   
13       3    1   39      1      5  31.2750          0           0   
14       3    0   14      0      0   7.8542          0           0   
15       2    0   55      0      0  16.0000          0           0   
16       3    1    2      4      1  29.1250          0           1   
17       2    1   29      0      0  13.0000          0           0   
18       3    0   31      1      0  18.0000          0           0   
19       3    0   29      0      0   7.2250          1           0   

    Southampton  A  B  C  D  E  F  G  T  U  Survived  
0             1  0  0  0  0  0  0  0  0  1         0  
1             0  0  0  1  0  0  0  0  0  0         1  
2             1  0  0  0  0  0  0  0  0  1         1  
3             1  0  0  1  0  0  0  0  0  0         1  
4             1  0  0  0  0  0  0  0  0  1         0  
5             0  0  0  0  0  0  0  0  0  1         0  
6             1  0  0  0  0  1  0  0  0  0         0  
7             1  0  0  0  0  0  0  0  0  1         0  
8             1  0  0  0  0  0  0  0  0  1         1  
9             0  0  0  0  0  0  0  0  0  1         1  
10            1  0  0  0  0  0  0  1  0  0         1  
11            1  0  0  1  0  0  0  0  0  0         1  
12            1  0  0  0  0  0  0  0  0  1         0  
13            1  0  0  0  0  0  0  0  0  1         0  
14            1  0  0  0  0  0  0  0  0  1         0  
15            1  0  0  0  0  0  0  0  0  1         1  
16            0  0  0  0  0  0  0  0  0  1         0  
17            1  0  0  0  0  0  0  0  0  1         1  
18            1  0  0  0  0  0  0  0  0  1         0  
19            0  0  0  0  0  0  0  0  0  1         1

In [27]:
df_preprocessed.to_csv('Titanic_preprocessed.csv', index = False)